cvat-ui/src/utils/opencv-wrapper/histogram-equalization.ts

```diff
@@ -0,0 +1,126 @@
+                channels.delete();
+                channels = null;
+                cv.equalizeHist(Y, equalizedY);
+                Y.delete();
+                channels = new cv.MatVector();
+                channels.push_back(equalizedY); equalizedY.delete();
+                channels.push_back(U); U.delete();
+                channels.push_back(V); V.delete();
+                cv.merge(channels, YUVDist);
+                cv.cvtColor(YUVDist, RGBDist, cv.COLOR_YUV2RGB, 0);
+                cv.cvtColor(RGBDist, RGBADist, cv.COLOR_RGB2RGBA, 0);
+                const arr = new Uint8ClampedArray(RGBADist.data, RGBADist.cols, RGBADist.rows);
+                const imgData = new ImageData(arr, src.width, src.height);
+                return createImageBitmap(imgData).then((bitmap:ImageBitmap) => {
+                    this.hashFrame(bitmap, frameNumber);
+                    return bitmap;
+                });
+            } catch (e) {
+                console.log('error in eq', e);
+                return undefined;
+            } finally {
+                if (matImage) matImage.delete();
+                if (channels) channels.delete();
+                RGBImage.delete();
+                YUVImage.delete();
+                RGBDist.delete();
+                YUVDist.delete();
+                RGBADist.delete();
+            }
+        } else {
+            this.currentUnequalized = src;
+            this.currentEqualizedNumber = frameNumber;
+            return hashedFrame;
+        }
+    }
+
+    private isHashed(frameNumber: number): ImageBitmap|undefined {
+        for (const elem of this.histHash) {
+            if (elem.frameNumber === frameNumber) {
+                elem.timestamp = window.performance.now();
+                return elem.bitmap;
+            }
+        }
+        return undefined;
+    }
+
+    public async restoreImage():Promise<ImageBitmap|undefined> {
+        if (this.currentUnequalized) {
+            return createImageBitmap(this.currentUnequalized);
+        }
+        return undefined;
+    }
+
+    private hashFrame(bitmap:ImageBitmap, frameNumber:number):void{
+        if (this.histHash.length >= this.bufferSize) {
+            const leastRecentlyUsed = this.histHash[0];
+            const currentTimestamp = window.performance.now();
+            let diff = currentTimestamp - leastRecentlyUsed.timestamp;
+            let leastIndex = 0;
+            for (let i = 1; i < this.histHash.length; i++) {
+                const currentDiff = currentTimestamp - this.histHash[i].timestamp;
+                if (currentDiff > diff) {
+                    diff = currentDiff;
+                    leastIndex = i;
+                }
+            }
+            this.histHash.splice(leastIndex, 1);
+        }
+        this.histHash.push({
+            bitmap,
+            frameNumber,
+            timestamp: window.performance.now(),
+        });
+    }
```

Comment:
I am not sure we need to buffer these frames here because they are buffered in cvat-data module. Can you check it?

Reply:
That buffer holds equalized frames, so i suppose we need it anyway